### PR TITLE
(#194) supports multiform

### LINKED ISSUE #194: Form: File upload
https://tutorialedge.net/golang/go-file-upload-tutorial/

### PR DESCRIPTION
**Pull request by dannypsnl**
close #194

### CODE REVIEW

**Comment by dannypsnl** [comment]
@shana0440 you can try: let the limit of the size of files can be configurable, I believe in the best case we can write it in tag, or worst we define it globally

File: filler.go [hunk 1]
```diff
@@ -177,6 +187,35 @@ func (f *formFiller) fill(ctx reflect.Value) error {
 	return nil
 }
 
+func (m *multiFormFiller) fill(ctx reflect.Value) error {
+	for k, idx := range m.multiFormParams {
+		// left shift 20 offset means MB
+		err := m.req.ParseMultipartForm(int64(m.limit[k]) << 20)
+		if err != nil {
+			return err
+		}
```
Comment (shana0440) [inline filler.go:196]: We should invoke `ParseMultipartForm` only once, if the user has many parameters in multi-form, it is weird that they have to set the limit to every parameter. Do you think if we take max memory from environment variables will be more reasonable? we can default it as 1MB if the user didn't specify like Nginx.

File: filler.go [hunk 2]
```diff
@@ -177,6 +187,35 @@ func (f *formFiller) fill(ctx reflect.Value) error {
...
+		file, _, err := m.req.FormFile(k)
+		if err != nil {
+			return err
+		}
```
Comment (shana0440) [inline filler.go:200]: multi-form not always is file, it might be value too.

File: filler.go (hunk 1 above)
Comment (dannypsnl) [reply filler.go:196]: Will users need different size limits for different kinds of files?

File: filler.go (hunk 2 above)
Comment (dannypsnl) [reply filler.go:200]: Then tag should be `multiformfile:"key"`? And distinct with `multiform`

Comment (dannypsnl) [reply filler.go:200]: Or just `file:"key"`?

File: filler.go (hunk 1 above)
Comment (shana0440) [reply filler.go:196]: From document, `ParseMultipartForm` only call once, the rest of the calls have no effect, we can only set the max memory that we expected request body will used, so user is not able to set the different size limit for different files.
https://golang.org/pkg/net/http/#Request.ParseMultipartForm

I think if user want one api can only receive 10MB file, another is 5MB, they might need to check the file size by themselves.

File: filler.go (hunk 2 above)
Comment (shana0440) [reply filler.go:200]: I prefer `multiform`, since I expected the tag should be Content-Type of http.

File: filler.go (hunk 1 above)
Comment (dannypsnl) [reply filler.go:196]: Make sense

File: filler.go (hunk 2 above)
Comment (dannypsnl) [reply filler.go:200]: Do you mean `multiform` for all cases? Probably we could use `multiform:"key" file:"yes"` to distinct them?

Comment (shana0440) [reply filler.go:200]: I think `multiform:"key" file:"yes"` is good.

**Comment by dannypsnl** [comment]
ci 爆了，起床再看www

**Comment by dannypsnl** [comment]
> ci 爆了，起床再看www

結果是小問題，修好了

**Comment by dannypsnl** [comment]
@shana0440 maybe we could use `OnLaunch` to set up multiple forms size-limit?

**Comment by shana0440** [comment]
@dannypsnl you mean [Launch()](https://github.com/dannypsnl/rocket/blob/65224de462b9c765db05bfb11c436360e83f55d6/rocket.go#L50)?
I think it's a good idea, we can add a parameter to receive the options, let the user decide how to control the server configuration.

**Comment by dannypsnl** [comment]
> @dannypsnl you mean [Launch()](https://github.com/dannypsnl/rocket/blob/65224de462b9c765db05bfb11c436360e83f55d6/rocket.go#L50)?
> I think it's a good idea, we can add a parameter to receive the options, let the user decide how to control the server configuration.

No, I mean `OnLaunch` of Fairing www. But then seems like this is a good idea to have configuration.

> add a parameter to receive the options

This can be considered too, which one you prefer? I think `OnLaunch` should be removed.

**Comment by dannypsnl** [comment]
on the other hand, changing API is that good.

**Comment by dannypsnl** [comment]
@shana0440 you can take a look at https://github.com/dannypsnl/rocket/pull/195/commits/061d98325eea2879146de902b49d250c18105670

File: rocket.go [hunk 3]
```diff
@@ -22,6 +22,9 @@ type Rocket struct {
 	// cache layer
 	defaultHandler reflect.Value
 	defaultResp    *response.Response
+
+	// SizeLimit decide the multiple forms value size
+	SizeLimit int64
```
Comment (shana0440) [inline rocket.go:27]: I prefer `MultiFormBodySizeLimit`, this should be more clear than `SizeLimit`.

**Review by shana0440 — COMMENTED** [review]
image

**Pull request merged**